src/apps/vpn/addons/state/addonstatebase.cpp

```diff
@@ -0,0 +1,122 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "addonstatebase.h"
+
+#include <QHash>
+#include <QJsonObject>
+
+#include "logger.h"
+
+namespace {
+Logger logger("AddonStateBase");
+}
+
+// static
+QJsonValue::Type AddonStateBase::typeToQJsonValueType(QString type) {
+  if (type == "boolean") {
+    return QJsonValue::Bool;
+  }
+
+  if (type == "number") {
+    return QJsonValue::Double;
+  }
+
+  if (type == "string") {
+    return QJsonValue::String;
+  }
+
+  return QJsonValue::Undefined;
+}
+
+// static
+QHash<QString, QJsonValue> AddonStateBase::parseManifest(
+    const QJsonObject& manifest) {
+  QHash<QString, QJsonValue> spec;
```

Comment:
Here you are creating a hash table, then you populate it, and then you return a copy of it as return value. What about if you receive a reference as param to avoid the extra copy.